Print environment variables for `cargo run` in extra verbose mode

Kobzol opened the pull request:
<!-- homu-ignore:start -->
### What does this PR try to resolve?

This PR changes `cargo run` so that it prints the environment variables that it sets for the executed binary in extra verbose mode (`cargo run -vv`). This is particularly useful when it sets `LD_LIBRARY_PATH` which is needed to actually run the binary. Because if users want to run the binary without `cargo`, they will need to know where should they point that variable. This is described in issue https://github.com/rust-lang/cargo/issues/4879.

Fixes: https://github.com/rust-lang/cargo/issues/4879

### How should we test and review this PR?

I tried to write a functional test, but I'm not sure how to check that `stderr` contains only a given substring, not a whole line.
<!-- homu-ignore:end -->

weihanglo commented on tests/testsuite/run.rs:
```diff
@@ -1416,6 +1416,23 @@ fn default_run_workspace() {
+        .file("src/main.rs", r#"fn main() {println!("run-a");}"#)
+        .build();
+
+    p.cargo("run -vv").with_stderr("LD_LIBRARY_PATH").run();
```
This contains two shortcomings:

* [Different platforms have different library paths](https://doc.rust-lang.org/nightly/cargo/reference/environment-variables.html#dynamic-library-paths). This won't pass on Windows and macOS.
* This matches both `rustc` and the final binary invocations. We can't tell whether the env is from which.

What about this? `[..]` for wildcard matches, and we match one env vars to verify the behavior.

```suggestion
    p.cargo("run -vv")
        .with_stderr(
            "\
[COMPILING] a v0.0.1 ([CWD])
[RUNNING] `[..]CARGO_MANIFEST_DIR=[CWD][..] rustc --crate-name a[..]`
[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
[RUNNING] `[..]CARGO_MANIFEST_DIR=[CWD][..] target/debug/a[EXE]`",
        )
        .run();
```

weihanglo reviewed (commented):
Looks good to me. Do you want to deal with `cargo test` as well in this pull request, or a follow-up?

(Can also leave it there. I am fine with that 😁)

arlosi commented on tests/testsuite/run.rs:
```diff
@@ -1416,6 +1416,23 @@ fn default_run_workspace() {
+        .file("src/main.rs", r#"fn main() {println!("run-a");}"#)
+        .build();
+
+    p.cargo("run -vv").with_stderr("LD_LIBRARY_PATH").run();
```
`LD_LIBRARY_PATH` won't be set on Windows. I'd recommend a different one like `CARGO_PKG_NAME` for the test case. 

You can use `[..]` as a wildcard for matching.

```suggestion
    p.cargo("run -vv")
        .with_stderr_contains("[RUNNING] [..]CARGO_PKG_NAME=a[..]")
        .run();
```

arlosi replied:
@weihanglo beat me to the review with the same comment 😊. I'll close this one.

Kobzol commented:
Thanks for the hints. I added support for `cargo test` too. I have to say, it would be really nice if `cargo` had auto-blessable tests with placeholders :laughing: Getting the text output right is not straightforward.

weihanglo commented:
I believe changing to `[RUNNING] `[..] rustdoc [..]--test [..]` would fix CI failures.

Kobzol commented:
Thanks for the hint! I fixed the two failing tests and used `[DOCTEST] foo` in the `test` test, as I noticed it was used in the failing tests. I also added a test for `bench`.

arlosi commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 4eac5a1fd152c62f7275896afb4379e39413e99e has been approved by `arlosi`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=arlosi 4eac5a1fd152c62f7275896afb4379e39413e99e -->
<!-- homu: {"type":"Approved","sha":"4eac5a1fd152c62f7275896afb4379e39413e99e","approver":"arlosi","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 4eac5a1fd152c62f7275896afb4379e39413e99e with merge 522e9fc4de0ca048d4a6226645687165bfdd8a25...
<!-- homu: {"type":"BuildStarted","head_sha":"4eac5a1fd152c62f7275896afb4379e39413e99e","merge_sha":"522e9fc4de0ca048d4a6226645687165bfdd8a25"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/5881875121/job/15952689647)
Approved by: arlosi
Pushing 522e9fc4de0ca048d4a6226645687165bfdd8a25 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"arlosi","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/5881875121/job/15952689647"},"merge_sha":"522e9fc4de0ca048d4a6226645687165bfdd8a25"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

arlosi commented:
@bors retry

bors commented:
:hourglass: Testing commit 4eac5a1fd152c62f7275896afb4379e39413e99e with merge 7b61184ad196e6b361a9a46e967899c618597557...
<!-- homu: {"type":"BuildStarted","head_sha":"4eac5a1fd152c62f7275896afb4379e39413e99e","merge_sha":"7b61184ad196e6b361a9a46e967899c618597557"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/5887163465/job/15967069914)
Approved by: arlosi
Pushing 7b61184ad196e6b361a9a46e967899c618597557 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"arlosi","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/5887163465/job/15967069914"},"merge_sha":"7b61184ad196e6b361a9a46e967899c618597557"} -->
